src/main/java/com/direwolf20/buildinggadgets/common/schema/MutableBlockData.java

```diff
@@ -0,0 +1,67 @@
+package com.direwolf20.buildinggadgets.common.schema;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.math.BlockPos;
+
+import javax.annotation.Nullable;
+import java.util.Objects;
+
+public final class MutableBlockData {
```

Comment:
We don't need a Mutable version of the BlockData. It should never change and I don't plan on it changing at any point. I'd prefer the implementation is Immutable by nature. Think of the BlockData as a fragment in time. Anything to be changed would happen on placement after fetching the data from here.

Reply:
You clearly don't have understood my intention for this (which is logical as I haven't added any comments yet). This exists purely to avoid Object allocations during iteration, I was considering allowing a forEach style instead of doing this, but decided against it, as we wouldn't be able to take advantage of all the already built API's around Iterable and Streams.

Reply:
Please use the lazy and simplest methods as long as it doesn't impede on readability. I'd rather something simple than not readable. Or at the very least, follow the MC standard by putting the Mutable instance of the class as a subclass.

Reply:
Upon this, again, please don't go over the top of memory savings.

Reply:
Remember the classical String concat in a loop test in contrast to using StringBuilder's. This would give (and gives the gc in the 1.14/1.15 impl) a similiar amount of work. If you look at the Iterator in the Template class you'll see that it uses only O(1) Object allocations.

Reply:
> Upon this, again, please don't go over the top of memory savings.

It's actually not a Memory saving as the footprint is in fact slightly worse for the Immutable variant, because itust load an additional object header. However as stated in my last comment it prevents the gc collector from being incredibly active when Templates are being consumed.

Reply:
> Or at the very least, follow the MC standard by putting the Mutable instance of the class as a subclass

@MichaelHillcox just because MC doesn't make it right, doesn't mean I'm going to make it wrong either.

A class can only be immutable if not only all it's members are unmodifiable, but it is also final or alternatively all members are either private or final. This is due the possibilitx of someoneelde being able to do exactly that: subclass it and make a mutable variant. The latter is horrible, because you can never be sure whether you have a truly immutable Object or something mutable in your hands and need to do explicit type checks to enforce this. Therefore it is invalid to make the mutable variant inherit from the immutable one. (If you don't believe me, ask some experienced people on the Forge Discord, I read HenryLoehenwind from EnderIO give a short Form of this to another modder on the discord about a year ago.) Mojang doing this is a pain for anyone who wants to store BlockPositions.
Whilst it would technically be valid to make the Immutable Variant inhert from the Mutable One, I kind of don't like the OO implication of the immutable variant also being a mutable instance... Therefore I think the guideline **Favor Delegation over Inheritance** of Effective Java applies here very well.

Reply:
Notice however that if we assume we make no programming errors (like not storing the mutable variant away without a copy) we'd not need the immutable variant...

Reply:
I'll agree on not having them as subclasses, I wasn't aware of that problem. Although, I still do not see the reason to have a mutable varient here. We really do not need to be that careful over the footprint thing system has as it's not something that is going to be running a lot. Hell, a good amount of players using Building Gadgets won't ever touch this. I'm trying to say to limit your thinking regarding memory and JVM handling and allow the code you write to be a bit more lazy for the sake of readability and useability. We don't need to worry about programming errors as we won't have other people using this specific system. We'll only be exposing the ability to supply safe NBT data to copy and even then I'm not sure how I want that implemented. I don't intend nor want other modders using this system. It should be bispokely created for our single use. I'm not saying to use the Immutable instance of the class in a mutable way, I'm saying to use it in an Immutable way but find a simpler solution to the problem you are trying to solve. The problem also being one that does not need to be solved as I'm requesting you dumb down the implemenation a bit. Always remember that Dire has to be able to work on this code. I would have already done a lot differently if Dire wasn't going to be using and contributing to this codebase. And no, telling dire to read up on some very specific use case or edge case to warrent your code isn't a valid argument, as dires knowledge grows we'll be able to expand upon the code we can right whilst always allowing dire to be able to contribute. Always remember, you're making a mod, not enterprise software, you want to make extreamly thoughtout and complex systems, Forge is always looking for more people to work on their systems.

Reply:
I'd draw the conclusion from this, that we drop the immutable variant and I call this `MitableTemplateData`, would that be ok for you? (going to comment on the rest later)

Reply:
How about just TemplateData with a java doc to explain how to use it?

Reply:
Well, ok.

Reply:
> your thinking regarding memory
It's about performance and not about memory...
> We'll only be exposing the ability to supply safe NBT data to copy and even then I'm not sure how I want that implemented
Just to make that clear: even if it wouldn't also contain Tile nbt, it would need to have some combined class in order be able to iterate it nicely. I could technically expose a getter accepting a position and returning the state and then allowing iteration over the blockstates, however this is imo a bit not nice, as we usually want entry set iteration and not key-set iteration. Addationally querying the blockstate adds more code and makes implementing something like the build scheduler slightly more complex, as it would not only need to have the iterator but also access to the Template... (The same goes of course for every other iteration: you'd need access to both...)

Reply:
Also not telling Dire to read up on something is imo the wrong way: if you don't know something you shouldn't just stop there and avoid it, but rather try to understand it when it may be needed. But I know that an argument about this is not going to go anywhere so, let's leave it with the result that we have different opinions.

Reply:
It's not opinions, it's an equation of time, I've learnt java in the abundance of time I've been luckily enough to have, dire unfortunately does not have that time to look into it all. Like when I work with your current template system, sure it works, sometimes, but it's a pain in the ass to work with, you need to know exactly how x system works and then how the y system works just to start looking at making something with the z system. Keep it simple stupid is effective the motto of the mod so try and keep it all simple. My attempt here is to try and restrain you in how you're implementing code so that myself and dire can work with it

Reply:
> It's about performance and not about memory...

ugh..

Reply:
> ugh..
Yes ok, that was worded badly - memory consumption is also a performance aspect. I meant that this specific implementation has been chosen in order to improve runtime performance, as frequent or heavy gc invocations (even with modern and pretty fast gc impls. - which aren't used by MC cause it's still on Java 8) will cause lag on low power machines.

Reply:
> how you're implementing code so that myself and dire can work with it
Whilst I accept that, I don't have to like it or agree with it...

Reply:
Maybe not but it's dires mod and I can't pull in code he won't be able to work with. So it's a must for this PR to be accepted

Reply:
As I said I accept it, but I neither like it nor agree with it.